### PR TITLE
feat: Object Storage

### PR DESCRIPTION
**Pull request by matoszz**
- adds new pkg/storage and associated storage backends
- adds GCS, AWS S3, and FS (filesystem) as initial backends
- creates basic operations for Save, Open, Delete for the respective backends with focus being on S3 for now
- adds struct tags for koanf and associated config structs (need to add enabled evals and such still)
- adds poorly written tests @golanglemonade will help me write better at some point ❤️ 

importing the package in a test file you can see basic operations function with credentials; additional tests, helpers, etc., to be added at a later date:
```
➜  datum git:(feat-storageinterfaces) ✗ go run schema/main.go
File saved to s3
attempting to read file...
File read from s3... file contents: (>_<)
➜  datum git:(feat-storageinterfaces) ✗ aws s3 ls s3://younohax0rzmybucket
2024-04-09 21:22:20          5 01HV30D91YAF7YYNK0DR681DA9
2024-04-09 22:19:46          5 01HV33PEB1QC2GQEQ5MA1NXKSS
2024-04-09 22:20:33          5 01HV33QW0YQ93GTYG3W6KKT3DC
2024-04-09 21:09:27          5 test
```

### CODE REVIEW

File: pkg/utils/storage/gcs/gcs.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+// Stat returns path metadata
+func (g *Storage) Stat(ctx context.Context, path string) (*gostorages.Stat, error) {
+	attrs, err := g.bucket.Object(path).Attrs(ctx)
+
+	if errors.Is(err, storage.ErrObjectNotExist) {
+		return nil, gostorages.ErrNotExist
+	} else if err != nil {
+		return nil, err
+	}
```
Comment (golanglemonade) [inline pkg/utils/storage/gcs/gcs.go:68]: nit: 

```suggestion
        if err != nil {
	    if errors.Is(err, storage.ErrObjectNotExist) {
		    return nil, gostorages.ErrNotExist
	    } 
	    return nil, err
	}
```

File: pkg/utils/storage/gcs/gcs.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	return errors.WithStack(g.bucket.Object(path).Delete(ctx))
+}
+
+// OpenWithStat opens path for reading with file stats
+func (g *Storage) OpenWithStat(ctx context.Context, path string) (io.ReadCloser, *gostorages.Stat, error) {
+	r, err := g.bucket.Object(path).NewReader(ctx)
+
+	if errors.Is(err, storage.ErrObjectNotExist) {
+		return nil, nil, gostorages.ErrNotExist
+	}
```
Comment (golanglemonade) [inline pkg/utils/storage/gcs/gcs.go:97]: this never checks for `err != nil` so you could have a bad time for errors other than ErrObjectNotExists when you try to get the `Attrs`

File: pkg/utils/storage/gcs/gcs_test.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+	if err != nil {
+		t.Errorf("unexpected error %v", err)
+	}
+
+	if stat.Size != 5 {
+		t.Errorf("expected size to be %d, got %d", 5, stat.Size)
+	}
+
+	assert.WithinDuration(t, before, now, time.Second)
```
Comment (golanglemonade) [inline pkg/utils/storage/gcs/gcs_test.go:54]: Same as on the other test, you use `assert` here but `if` statements other places

**Review by golanglemonade — COMMENTED** [review]
This is great! Just a few comments, (and patterns that were repeated so I stopped saying "same" at somepoint haha)

File: pkg/utils/storage/fs/fs.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+// Stat returns path metadata
+func (fs *Storage) Stat(ctx context.Context, path string) (*storage.Stat, error) {
+	fi, err := os.Stat(fs.abs(path))
+	if os.IsNotExist(err) {
+		return nil, storage.ErrNotExist
+	} else if err != nil {
+		return nil, errors.WithStack(err)
+	}
```
[the remark this answers is not in the document]
Comment (matoszz) [reply pkg/utils/storage/fs/fs.go:65]: @golanglemonade did you actually mean: 
```
	fi, err := os.Stat(fs.abs(path))
	if err != nil {
		if errors.Is(err, storage.ErrNotExist) {
			return nil, storage.ErrNotExist
		}
		
		return nil, errors.WithStack(err)
	}
```
? `fs.ErrNotExist` isn't defined

**Review by golanglemonade — APPROVED** [review]

**Pull request merged**